src/RamExpression.h

```diff
@@ -40,12 +40,52 @@ class RamExpression : public RamNode {
     RamExpression* clone() const override = 0;
 };
 
-// TODO (azreika): create a common abstract base class for RAM operators
+/**
+ * Abstract class for an operator/functor
+ */
+class RamAbstractOperator : public RamExpression {
+public:
+    RamAbstractOperator(std::vector<std::unique_ptr<RamExpression>> args) : arguments(std::move(args)) {}
+
+    /** Get argument values */
+    std::vector<RamExpression*> getArguments() const {
+        return toPtrVector(arguments);
+    }
+
+    /** Get i-th argument value */
+    const RamExpression& getArgument(size_t i) const {
+        assert(i >= 0 && i < arguments.size() && "argument index out of bounds");
+        return *arguments[i];
+    }
+
+    /** Get number of arguments */
+    size_t getArgCount() const {
+        return arguments.size();
+    }
+
+    std::vector<const RamNode*> getChildNodes() const override {
+        std::vector<const RamNode*> res;
+        for (const auto& cur : arguments) {
+            res.push_back(cur.get());
+        }
+        return res;
+    }
+
+    void apply(const RamNodeMapper& map) override {
+        for (auto& arg : arguments) {
+            arg = map(std::move(arg));
+        }
+    }
+
+protected:
+    /** Arguments of user defined operator */
+    std::vector<std::unique_ptr<RamExpression>> arguments;
+};
```

Comment:
Please, can you add an equal method here that checks for equivalence of arguments?